### PR TITLE
support return CR deposit coin transaction

### PR DESCRIPTION
**Pull request by RainFallsSilent**

### CODE REVIEW

File: blockchain/txvalidator.go [hunk 1]
```diff
@@ -1551,6 +1564,62 @@ func (b *BlockChain) checkReturnDepositCoinTransaction(txn *Transaction,
...
+			return err
+		}
+		programHash := ct.ToProgramHash()
+		// todo get candiadte form not voting period state.
+		c := b.crCommittee.GetState().GetCandidateByDID(*programHash)
+		if c == nil {
+			return errors.New("signer must be CR candidate")
+		}
+
+		if isInVotingPeriod(currentHeight) {
```
Comment (Houshoupei84) [inline blockchain/txvalidator.go]: this if and else move out for loop maybe better.

File: cr/state/state.go [hunk 2]
```diff
@@ -305,60 +353,78 @@ func (s *State) updateCandidateInfo(origin *payload.CRInfo, update *payload.CRIn
...
-				s.Votes[op.ReferKey()] = output
-				s.processVoteOutput(output, height)
+			if output.Type != types.OTVote {
+				continue
+			}
+			p, ok := output.Payload.(*outputpayload.VoteOutput)
+			if !ok {
+				continue
+			}
+			if p.Version == outputpayload.VoteProducerAndCRVersion {
```
Comment (raindust) [inline cr/state/state.go]: should be >=

Comment (RainFallsSilent) [reply cr/state/state.go]: fixed

File: blockchain/txvalidator.go [hunk 3]
```diff
@@ -1551,6 +1562,61 @@ func (b *BlockChain) checkReturnDepositCoinTransaction(txn *Transaction,
...
+
+	var penalty common.Fixed64
+	for _, program := range txn.Programs {
+		// Get candidate from code.
+		ct, err := contract.CreateCRDIDContractByCode(program.Code)
+		if err != nil {
+			return err
+		}
+		programHash := ct.ToProgramHash()
+		// todo get candiadte form not voting period state.
```
Comment (yz89) [inline blockchain/txvalidator.go]: candidate typo

**Comment by peterwillcn** [comment]
@RainFallsSilent  Check addlicense
core/types/payload/unregistercr_test.go
test/transaction/unregister_cr_tx.lua

File: blockchain/txvalidator.go (hunk 3 above)
Comment (RainFallsSilent) [reply blockchain/txvalidator.go]: fixed

File: cr/state/state.go [hunk 4]
```diff
@@ -305,7 +354,23 @@ func (s *State) updateCandidateInfo(origin *payload.CRInfo, update *payload.CRIn
...
+			if !ok {
+				continue
+			}
+			if p.Version < outputpayload.VoteProducerAndCRVersion {
+				continue
+			}
+			var exist bool
+			for _, content := range p.Contents {
+				if content.VoteType == outputpayload.CRC {
+					exist = true
```
Comment (Houshoupei84) [inline cr/state/state.go]: maybe add break is more better ?

File: cr/state/state.go [hunk 5]
```diff
@@ -108,6 +108,51 @@ func (s *State) ExistCandidateByNickname(nickname string) bool {
...
+				for _, content := range p.Contents {
+					if content.VoteType == outputpayload.CRC {
+						return true
+					}
+				}
+			}
+		}
+	}
+
+	// Cancel votes.
```
Comment (Houshoupei84) [inline cr/state/state.go]: only this section need lock

Comment (RainFallsSilent) [reply cr/state/state.go]: fixed

File: cr/state/state.go (hunk 4 above)
Comment (RainFallsSilent) [reply cr/state/state.go]: fixed

**Pull request merged**